api/metric/value.go

```diff
@@ -0,0 +1,193 @@
+// Copyright 2019, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metric
+
+import (
+	"fmt"
+	"math"
+)
+
+type valueKind int8
+
+const (
+	int64Kind valueKind = iota
+	float64Kind
+)
+
+// MeasurementValue represents either an integral or a floating point
+// value of a measurement. Empty MeasurementValue is treated as
+// integral zero.
+type MeasurementValue struct {
+	raw  uint64
+	kind valueKind
+}
+
+// MeasurementValueVisitor is an interface used for visitation of
+// MeasurementValues.
+type MeasurementValueVisitor interface {
```

Comment:
I have a vague feeling of worry over the interface conversion that will happen here, but I don't quite understood how it is going to be used. I'll study the code more.

Reply:
I'm not sure what do you mean about the interface conversion, but I'm curious about what do you honestly think about it (like overengineered? :) ). In the end, if more people will say that the float/int distinction is not necessary, then this can be made simpler (by dropping it?) with only storing float64 and doing casts for int64.

I also had some continuation of this work with the MutatingValueVisitor interface (which maybe could even replace the ValueVisitor interface), which allows you to modify the MeasurementValue in place (so to implement setting and adding values, atomically or not), [the branch/commit is here](https://github.com/kinvolk/opentelemetry-go/commit/61a4fc4187a64afa1df90088321f058af255d994), but I haven't yet pushed that - I'm not really sure if SDK is going to use that.

Reply:
I guess I think it's a bit overengineered.
In Go-1.4 they removed a really nice optimization, which allowed interface values to store single-word values.
Ever since then if you store a non-pointer in an interface, it's copied-by-value (and stored as a pointer internally). So this code:

```
func (v MeasurementValue) CoerceToInt64() int64 {
	var i int64
	c := int64Coercer{
		i: &i,
	}
	v.Visit(c)
	return i
}
```
I believe performs at least one allocation, maybe two, unless the compiler has gotten a lot better.  One is `&i`, and the other is the interface conversion I speak of.

For the metrics SDK code path, I'd like to avoid allocations, but I also want to use those coercion methods. They can be implemented without those allocations.  One solution would be to expose the value kind, then I could check the kind and do the coercion myself.

I also found a use for testing if the value is negative without needing to know its type. Just some thoughts.

Overall I guess the visitor pattern is over-engineered for the likely use-cases.

Reply:
I had a look at #172. So I reworked the value stuff. It's in [my branch](https://github.com/kinvolk/opentelemetry-go/tree/krnowak/metrics), I haven't yet pushed it here.

I have changed it to basically `type MeasurementValue uint64`. The type has `AsInt64() int64` and `AsFloat64() float64` functions and some helper methods for comparisons (`RawCompare(other uint64, kind ValueKind)`, `Float64Compare(float64)`, `Int64Compare(int64)`, `IsPositive(ValueKind)`, `IsZero(ValueKind)`, `IsNegative(ValueKind)`). You can also get raw uint64 (either with cast or with the `AsRaw() uint64` method) and a pointer to raw (either some other cast or with `AsRawPtr() *uint64`). So hopefully this will be closer to your/SDK needs.

Reply:
I have taken your sdk-metrics branch and rebased it on top of my current work to see how the code would look like now, you can see it here: https://github.com/kinvolk/opentelemetry-go/tree/krnowak/sdk-metrics